File: C# Examples/14 - 15 Creating Add-ins/MyMacroAddIn - Part2/MyMacroAddIn/UserPMPage.cs
using System;
using SolidWorks.Interop.sldworks;
using SolidWorks.Interop.swpublished;
using SolidWorks.Interop.swconst;

namespace MyMacroAddIn
{
    public class UserPMPage
    {
        //Local Objects
        public IPropertyManagerPage2 swPropertyPage = null;
        PMPHandler handler = null;
        ISldWorks iSwApp = null;
        SwAddin userAddin = null;

        #region Property Manager Page Controls
        //Groups
        IPropertyManagerPageGroup group2;

        //Controls
        public IPropertyManagerPageTextbox textbox1;
        IPropertyManagerPageSelectionbox selection1;
        IPropertyManagerPageButton button1;

        //Control IDs
        public const int group2ID = 1;

        public const int textbox1ID = 2;
        public const int selection1ID = 8;
        public const int buttonID1 = 13;

        #endregion

        public UserPMPage(SwAddin addin)
        {
            userAddin = addin;
            if (userAddin != null)
            {
                iSwApp = (ISldWorks)userAddin.SwApp;
                CreatePropertyManagerPage();
            }
            else
            {
                System.Windows.Forms.MessageBox.Show("SwAddin not set.");
            }
        }


        protected void CreatePropertyManagerPage()
        {
            int errors = -1;
            int options = (int)swPropertyManagerPageOptions_e.swPropertyManagerOptions_OkayButton |
                (int)swPropertyManagerPageOptions_e.swPropertyManagerOptions_CancelButton;

            handler = new PMPHandler(userAddin, this);
            swPropertyPage = (IPropertyManagerPage2)iSwApp.CreatePropertyManagerPage("Add Component", options, handler, ref errors);
            if (swPropertyPage != null && errors == (int)swPropertyManagerPageStatus_e.swPropertyManagerPage_Okay)
            {
                try
                {
                    AddControls();
                }
                catch (Exception e)
                {
                    iSwApp.SendMsgToUser2(e.Message, 0, 0);
                }
            }
        }


        //Controls are displayed on the page top to bottom in the order 
        //in which they are added to the object.
        protected void AddControls()
        {
            short controlType = -1;
            short align = -1;
            int options = -1;
            bool retval;

            //Add Message
            string message = "Select one or more circular edges where the component will be mated. "
                + "Browse to a fastener to insert, then click OK.";
;
            retval = swPropertyPage.SetMessage3(message,
                                            (int)swPropertyManagerPageMessageVisibility.swImportantMessageBox,
                                            (int)swPropertyManagerPageMessageExpanded.swMessageBoxExpand,
                                            "Message");

            //Add the groups
            options = (int)swAddGroupBoxOptions_e.swGroupBoxOptions_Expanded | 
                      (int)swAddGroupBoxOptions_e.swGroupBoxOptions_Visible;

            group2 = (IPropertyManagerPageGroup)swPropertyPage.AddGroupBox(group2ID, "Add Component", options);

            //Add the controls to group1

            //Add controls to group2
            //selection1
            controlType = (int)swPropertyManagerPageControlType_e.swControlType_Selectionbox;
            align = (int)swPropertyManagerPageControlLeftAlign_e.swControlAlign_LeftEdge;
            options = (int)swAddControlOptions_e.swControlOptions_Enabled |
                      (int)swAddControlOptions_e.swControlOptions_Visible;

            selection1 = (IPropertyManagerPageSelectionbox)group2.AddControl(selection1ID, controlType, "Select Circular Edges", 
                align, options, "Select circular edges on flat faces where components will be added");
            if (selection1 != null)
            {
                int[] filter = { (int)swSelectType_e.swSelEDGES };
                selection1.Height = 50;
                selection1.SetSelectionFilters(filter);
            }

            // Button
            controlType = (int)swPropertyManagerPageControlType_e.swControlType_Button;
            align = (int)swPropertyManagerPageControlLeftAlign_e.swControlAlign_LeftEdge;
            options = (int)swAddControlOptions_e.swControlOptions_Enabled |
                    (int)swAddControlOptions_e.swControlOptions_Visible;

            button1 = (IPropertyManagerPageButton)group2.AddControl2(buttonID1, controlType, "Browse...", 
                align, options, "Browse to a part");

            //label
            group2.AddControl(100, 1, "File path", 1, 3, "");

            //textbox1
            controlType = (int)swPropertyManagerPageControlType_e.swControlType_Textbox;
            align = (int)swPropertyManagerPageControlLeftAlign_e.swControlAlign_LeftEdge;
            options = (int)swAddControlOptions_e.swControlOptions_Enabled |
                      (int)swAddControlOptions_e.swControlOptions_Visible;

            textbox1 = (IPropertyManagerPageTextbox)group2.AddControl(textbox1ID, controlType, "File path", 
                align, options, "File path");

        }

        public void Show()
        {
            if (swPropertyPage != null)
            {
                swPropertyPage.Show();
            }
        }
    }
}

File: 14 - 15 Creating Add-ins/MyMacroAddIn/MyMacroAddIn/UserPMPage.vb
Imports SolidWorks.Interop.sldworks
Imports SolidWorks.Interop.swconst
Imports SolidWorks.Interop.swpublished

Public Class UserPMPage
    Dim iSwApp As SldWorks
    Dim userAddin As SwAddin
    Dim handler As PMPageHandler
    Friend PropMgrPage As PropertyManagerPage2

#Region "Property Manager Page Controls"
    'Groups
    Dim group2 As PropertyManagerPageGroup

    'Controls
    Dim selection1 As PropertyManagerPageSelectionbox
    Dim button1 As PropertyManagerPageButton
    Friend text1 As PropertyManagerPageTextbox

    'Control IDs
    Dim group2ID As Integer = 1
    Dim selection1ID As Integer = 7
    Friend buttonID1 As Integer = 12
    Dim textID1 As Integer = 14

#End Region

    Sub Init(ByVal sw As SldWorks, ByVal addin As SwAddin)
        iSwApp = sw
        userAddin = addin
        CreatePage()
        AddControls()
    End Sub

    Sub Show()
        PropMgrPage.Show()
    End Sub

    Sub CreatePage()
        handler = New PMPageHandler()
        handler.Init(iSwApp, userAddin, Me)
        Dim options As Integer
        Dim errors As Integer
        options = swPropertyManagerPageOptions_e.swPropertyManagerOptions_OkayButton + swPropertyManagerPageOptions_e.swPropertyManagerOptions_CancelButton
        PropMgrPage = iSwApp.CreatePropertyManagerPage("Add Component", options, handler, errors)
    End Sub

    Sub AddControls()
        Dim options As Integer
        Dim leftAlign As Integer
        Dim controlType As Integer
        Dim retval As Boolean

        ' Add Message
        Dim message As String
        message = "Select one or more circular edges where the component will be mated. " _
      & "Browse to a fastener to insert, then click OK."
        retval = PropMgrPage.SetMessage3(message,
                                      swPropertyManagerPageMessageVisibility.swImportantMessageBox,
                                      swPropertyManagerPageMessageExpanded.swMessageBoxExpand,
                                      "Message")

        'Add Groups
        options = swAddGroupBoxOptions_e.swGroupBoxOptions_Checkbox + swAddGroupBoxOptions_e.swGroupBoxOptions_Visible
        group2 = PropMgrPage.AddGroupBox(group2ID, "Add Component", options)


        'Add Controls to Group2
        'Selection1
        controlType = swPropertyManagerPageControlType_e.swControlType_Selectionbox
        leftAlign = swPropertyManagerPageControlLeftAlign_e.swControlAlign_LeftEdge
        options = swAddControlOptions_e.swControlOptions_Enabled + swAddControlOptions_e.swControlOptions_Visible
        selection1 = group2.AddControl(selection1ID, controlType, "Select Circular Edges", leftAlign, options,
                                       "Select circular edges on flat faces where components will be added")
        If Not selection1 Is Nothing Then
            Dim filter() As Integer = New Integer() {swSelectType_e.swSelEDGES}
            selection1.Height = 50
            selection1.SetSelectionFilters(filter)
        End If

        'Label
        group2.AddControl(100, 1, "File path", 1, 3, "")

        'Textbox1
        controlType = swPropertyManagerPageControlType_e.swControlType_Textbox
        leftAlign = swPropertyManagerPageControlLeftAlign_e.swControlAlign_LeftEdge
        options = swAddControlOptions_e.swControlOptions_Enabled +
                    swAddControlOptions_e.swControlOptions_Visible
        text1 = group2.AddControl2(textID1, controlType, "File path",
                    leftAlign, options, "File path")

        'Button
        controlType = swPropertyManagerPageControlType_e.swControlType_Button
        leftAlign = swPropertyManagerPageControlLeftAlign_e.swControlAlign_DoubleIndent
        options = swAddControlOptions_e.swControlOptions_Enabled +
                    swAddControlOptions_e.swControlOptions_Visible
        button1 = group2.AddControl2(buttonID1, controlType, "Browse...",
                    leftAlign, options, "Browse to a part")

    End Sub

End Class

